File: worker/src/attestation.ts
import { createHash, createHmac, timingSafeEqual } from "node:crypto";
import { readFile } from "node:fs/promises";
import { resolve } from "node:path";

export type AttestationEnvelope = {
  document: {
    enclaveImage: string;
    enclaveVersion: string;
    pcrs: Record<string, string>;
    publicKey: string;
    issuedAt: string;
    expiresAt: string;
  };
  signature: string;
};

export type AttestationVerification = {
  ok: boolean;
  reasons: string[];
};

const sha256 = (value: string) => createHash("sha256").update(value).digest("hex");

export const expectedImage = "sovereign-vault-python-fastapi";
export const expectedPcrs = {
  PCR0: sha256("vault-fastapi-image:2026.05.local").slice(0, 64),
  PCR1: sha256("vault-runtime-python3.12").slice(0, 64),
  PCR2: sha256("sovereign-vault-policy-bundle").slice(0, 64)
};

export function signDocument(
  document: AttestationEnvelope["document"],
  secret = process.env.ATTESTATION_SECRET ?? "local-dev-attestation-secret"
): string {
  return createHmac("sha256", secret).update(JSON.stringify(document)).digest("hex");
}

export function verifyAttestation(
  envelope: AttestationEnvelope,
  secret = process.env.ATTESTATION_SECRET ?? "local-dev-attestation-secret"
): AttestationVerification {
  const reasons: string[] = [];
  const expectedSignature = signDocument(envelope.document, secret);
  const actual = Buffer.from(envelope.signature, "hex");
  const expected = Buffer.from(expectedSignature, "hex");

  if (actual.length !== expected.length || !timingSafeEqual(actual, expected)) {
    reasons.push("Worker [Attestation]: signature mismatch; refusing to send audit task");
  }

  if (envelope.document.enclaveImage !== expectedImage) {
    reasons.push("Worker [Attestation]: enclave image identity mismatch");
  }

  for (const [pcr, expectedPcr] of Object.entries(expectedPcrs)) {
    if (envelope.document.pcrs[pcr] !== expectedPcr) {
      reasons.push(`Worker [Attestation]: ${pcr} measurement mismatch`);
    }
  }

  if (new Date(envelope.document.expiresAt).getTime() <= Date.now()) {
    reasons.push("Worker [Attestation]: identity proof expired");
  }

  return { ok: reasons.length === 0, reasons };
}

export async function loadAttestation(path = defaultAttestationPath()): Promise<AttestationEnvelope> {
  const content = await readFile(path, "utf8");
  return JSON.parse(content) as AttestationEnvelope;
}

export function defaultAttestationPath(): string {
  return resolve(process.env.ATTESTATION_DOC_PATH ?? "scripts/attestation/enclave-attestation.json");
}

File: vault/app/attestation.py
from __future__ import annotations

import hashlib
import hmac
import json
import os
from dataclasses import dataclass
from datetime import datetime, timezone

from .schemas import AttestationEnvelope


def _sha256(value: str) -> str:
    return hashlib.sha256(value.encode("utf-8")).hexdigest()


EXPECTED_IMAGE = "sovereign-vault-python-fastapi"
EXPECTED_PCRS = {
    "PCR0": _sha256("vault-fastapi-image:2026.05.local")[:64],
    "PCR1": _sha256("vault-runtime-python3.12")[:64],
    "PCR2": _sha256("sovereign-vault-policy-bundle")[:64],
}


@dataclass(frozen=True)
class AttestationResult:
    ok: bool
    reasons: list[str]


def verify_attestation(envelope: AttestationEnvelope) -> AttestationResult:
    secret = os.getenv("ATTESTATION_SECRET", "local-dev-attestation-secret")
    document = envelope.document.model_dump()
    canonical = json.dumps(document, separators=(",", ":"))
    expected_signature = hmac.new(
        secret.encode("utf-8"), canonical.encode("utf-8"), hashlib.sha256
    ).hexdigest()

    reasons: list[str] = []

    if not hmac.compare_digest(expected_signature, envelope.signature):
        reasons.append("Vault [Attestation]: signature mismatch; refusing audit task")

    if document["enclaveImage"] != EXPECTED_IMAGE:
        reasons.append("Vault [Attestation]: enclave image identity mismatch")

    for pcr, expected in EXPECTED_PCRS.items():
        if document["pcrs"].get(pcr) != expected:
            reasons.append(f"Vault [Attestation]: {pcr} measurement mismatch")

    expires_at = _parse_iso(document["expiresAt"])
    if expires_at <= datetime.now(timezone.utc):
        reasons.append("Vault [Attestation]: identity proof expired")

    return AttestationResult(ok=not reasons, reasons=reasons)


def _parse_iso(value: str) -> datetime:
    normalized = value.replace("Z", "+00:00")
    parsed = datetime.fromisoformat(normalized)
    if parsed.tzinfo is None:
        return parsed.replace(tzinfo=timezone.utc)
    return parsed.astimezone(timezone.utc)
